xds/src/main/java/io/grpc/xds/RbacFilter.java

```diff
@@ -0,0 +1,308 @@
+    List<GrpcAuthorizationEngine.PolicyMatcher> policyMatchers = new ArrayList<>();
+    for (Map.Entry<String, Policy> entry: policyMap.entrySet()) {
+      try {
+        Policy policy = entry.getValue();
+        if (policy.hasCondition() || policy.hasCheckedCondition()) {
+          return ConfigOrError.fromError(
+                  "Policy.condition and Policy.checked_condition must not set.");
+        }
+        policyMatchers.add(new PolicyMatcher(entry.getKey(),
+                parsePermissionList(policy.getPermissionsList()),
+                parsePrincipalList(policy.getPrincipalsList())));
+      } catch (Exception e) {
+        logger.log(Level.FINE, "Unexpected error parsing policy {0}", new Object[]{entry});
+        return ConfigOrError.fromError("Encountered error parsing policy: " + e);
+      }
+    }
+    return ConfigOrError.fromConfig(RbacConfig.create(new AuthConfig(policyMatchers, authAction)));
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfigOverride(Message rawProtoMessage) {
+    return parseFilterConfig(rawProtoMessage);
+  }
+
+  @Override
+  public ServerInterceptor buildServerInterceptor(FilterConfig config,
+                                                  @Nullable FilterConfig overrideConfig) {
+    checkNotNull(config, "config");
+    if (overrideConfig != null) {
+      config = overrideConfig;
+    }
+    AuthConfig authConfig = ((RbacConfig) config).authConfig();
+    return generateAuthorizationInterceptor(authConfig);
+  }
+
+  private ServerInterceptor generateAuthorizationInterceptor(AuthConfig config) {
+    final GrpcAuthorizationEngine authEngine = new GrpcAuthorizationEngine(config);
+    return new ServerInterceptor() {
+        @Override
+        public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+                final ServerCall<ReqT, RespT> call,
+                final Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+          AuthDecision authResult = authEngine.evaluate(headers, call);
+          logger.log(Level.FINE,
+                  "Authorization result for serverCall {0}: {1}, matching policy: {2}.",
+                  new Object[]{call, authResult.decision(), authResult.matchingPolicyName()});
+          if (GrpcAuthorizationEngine.Action.DENY.equals(authResult.decision())) {
+            Status status = Status.UNAUTHENTICATED.withDescription(
+                    "Access Denied, matching policy: " + authResult.matchingPolicyName());
+            call.close(status, new Metadata());
+            return new NoopListener<>();
+          }
+          return next.startCall(call, headers);
+        }
+    };
+  }
+
+  private static OrMatcher parsePermissionList(List<Permission> permissions) {
+    List<Matcher> anyMatch = new ArrayList<>();
+    for (Permission permission : permissions) {
+      anyMatch.add(parsePermission(permission));
+    }
+    return new OrMatcher(anyMatch);
```

Comment:
I'd generally wrap the list with `Collections.unmodifiableList()` to avoid leaking the mutability. May not be a big deal, as the API is not user-facing...

Reply:
yea everything is inline...also orMatcher does a copy immediately so don't see this local ArrayList hurts anyone.